ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c

```diff
@@ -37,45 +39,131 @@
 void init_filter_nnfw (void) __attribute__ ((constructor));
 void fini_filter_nnfw (void) __attribute__ ((destructor));
 
+typedef struct
+{
+  nnfw_tensorinfo i_in;
+  nnfw_tensorinfo i_out;
+  nnfw_session *session;
+  gchar *model_path;
+} nnfw_pdata;
+
+static void nnfw_close (const GstTensorFilterProperties * prop,
+    void **private_data);
+
 /**
  * @brief The standard tensor_filter callback
  */
-static int nnfw_open (const GstTensorFilterProperties * prop,
-    void **private_data)
+static int
+nnfw_open (const GstTensorFilterProperties * prop, void **private_data)
 {
-  return 0;
+  NNFW_STATUS status;
+  int err = 0;
+  nnfw_pdata *pdata;
+
+  if (*private_data != NULL) {
+    pdata = *private_data;
+    if (strcmp (prop->model_file, pdata->model_path)) {
+      nnfw_close (prop, private_data);  /* "reopen" */
+    } else {
+      return 1;
+    }
+  }
+
+  pdata = g_new0 (nnfw_pdata, 1);
+  if (pdata == NULL)
+    return -ENOMEM;
+
+  *private_data = (void *) pdata;
+
+  status = nnfw_create_session (&pdata->session);
+  if (status != NNFW_STATUS_NO_ERROR) {
+    err = -EINVAL;
+    g_printerr ("Cannot create nnfw-runtime session");
+    goto unalloc_exit;
+  }
+
+  status = nnfw_load_model_from_file (pdata->session, prop->model_file);
+  if (status != NNFW_STATUS_NO_ERROR) {
+    err = -EINVAL;
+    g_printerr ("Cannot load the model file: %s", prop->model_file);
+    goto unalloc_exit;
+  }
+
+  status = nnfw_prepare (pdata->session);
+  if (status != NNFW_STATUS_NO_ERROR) {
+    err = -EINVAL;
+    g_printerr ("nnfw-runtime cannot prepare the session for %s",
+        prop->model_file);
+    goto unalloc_exit;
+  }
+
+  pdata->model_path = g_strdup (prop->model_file);
+
+  goto exit;
+
+unalloc_exit:
+  g_free (pdata);
```

Comment:
add `nnfw_close (prop, private_data)`?

Reply:
The final states of variables are identical with nnfw_close() call; however, if we call nnfw_close(), it will be a inappropriate call for nnfw_close(), which will incur error (```"nnfw_close called without proper nnfw_open"```) by nnfw_close(), because nnfw_close assumes that nnfw_open is ended successfully.

Reply:
Then, I think line 88 and 95 should call `nnfw_close_session()` to cleanup already opened session.